### ISSUE TITLE
Feature: asynchronous processing

### ISSUE THREAD

**Issue by pascal08** [cloudcreativity/laravel-json-api#76]
I have a JsonApi endpoint `/prices`. When I submit a post request to this endpoint I would like to dispatch a job that is going to grab prices from a couple of external sources. I would like to return a response immediately that contains a reference to the dispatched job, like this: http://jsonapi.org/recommendations/#asynchronous-processing

This would be a high-level description of how I would solve this:
1. Create a JsonApi endpoint for fetching the `job` resources. 
2. Override the `create` method on `PricesController` with some code to dispatch the price fetching job
3. Return a response containing a reference url to the `job` resource.

This would return something like this:
```
HTTP/1.1 202 Accepted
Content-Type: application/vnd.api+json
Content-Location: https://example.com/jobs/1234

{
  "data": {
    "type": "jobs",
    "id": "1234",
    "attributes": {
      "status": "Pending request, waiting other process"
    },
    "links": {
      "self": "/jobs/1234"
    }
  }
}
```

When the job is processed this endpoint returns:
```
HTTP/1.1 303 See other
Content-Type: application/vnd.api+json
Location: https://example.com/prices/4321
``` 

I think it would be nice if this package supports this recommendation. What do you think? I'd be happy to help integrating this.

**Comment by lindyhopchris** [comment]
Absolutely fantastic idea... I definitely have multiple use-cases for this, and it would be a great compliment to Laravel's queue system. It would also be good if the job result was broadcasted.

So yes, this should definitely be in this package. The only thing to say is I've got quite a long list of stuff to do for v1.0 (see #60) and I need to prioritise those things. So I think this might be a v1.1 addition. Your help integrating it would definitely be appreciated because I won't get to it until I've finished the v1.0 stuff.

Here's my initial thoughts on how it could be implemented (nothing set in stone)....

1. Have a `ClientDispatchable` trait that is applied to jobs. This adds a `clientDispatch` static method, that wraps the `Illuminate\Foundation\Bus\Dispatchable::dispatch` method. This wrapper:
  - Applies a UUID to a property on the job
  - Saves its details to a database table, which can also store any "client-friendly" details about the job, i.e. stuff we can return in the request to the jobs endpoint.
  - Calls the standard Laravel dispatch method so that it gets added to the regular job queue.
  - Returns the JSON API database model, which can then be serialized by the controller for the 202 response. (All the controller would need to do is return this model instead of whatever normal model that it returns - the encoding stuff would all work if the schema for the `queue-jobs` resource is registered).

2.  Bind to the Laravel job events so that the database model of the job can be updated (using the UUID stored on the job object). We can provide a standard observer class that people can use if the want to broadcast these updates.

3. Have a controller that can respond to the `queue-jobs` resource requests accordingly.

Some extra thoughts:

1. Would need an Artisan command to generate the migration for the table.
2. Would need an Artisan command to clean out old job data from the table (for people to use via their cron schedule).
3. The resource would need to be called `queue-jobs` because `jobs` I think is too generic and might conflict.
4. The stored data would need to record which JSON API it relates to, so that for multi-API applications it knows which one the job relates to.

This is definitely an awesome idea! Thanks for the suggestion.

**Comment by pascal08** [comment]
Awesome! Let's see what I can do (I probably need some more guidance along the way). I'll keep you up-to-date.

**Comment by lindyhopchris** [comment]
Yep, no problem at all... definitely fire questions my way. 

The extra thought I've had on it since is actually the database model could be attached to the job rather than the UUID.... maybe on a `resource` property. That way the `SerializesModels` trait will handle pulling it out of the DB when it is deserializing the job.

**Comment by pascal08** [comment]
You mean a extra column on the `jobs` table (the one from Laravel) containing the serialized resource model?

**Comment by lindyhopchris** [comment]
Yep, definitely mean the serializes model trait rather than an extra column on the jobs table. We need to keep our data in a separate table.

**Comment by lindyhopchris** [comment]
One extra thought on this. The spec has the job route as `GET /photos/queue-jobs/5234`. We'd want to keep that nesting so that jobs are related to a specific resource type. That will be important for e.g. authorizing inbound requests.

**Comment by pascal08** [comment]
I agree. Still just one table `queue-jobs` for all resource related jobs, right?

**Comment by lindyhopchris** [comment]
Yep!

**Comment by plunkettscott** [comment]
Any update on this as being included within the package?

**Comment by lindyhopchris** [comment]
I still think this would be good to add (which is why the issue is still open) but it's not a priority at the moment as I'm using available time to work on the remaining `1.0.0` issues.

**Comment by lindyhopchris** [comment]
@plunkettscott it's possible this might be dealt with in the next month or so, as we're looking at using the async recommendation for a new API that we're currently building.
